python/cudf/cudf/pylibcudf_tests/conftest.py

```diff
@@ -20,31 +23,124 @@
-        pa.float64(),
-        pa.uint64(),
-    ],
+    params=NUMERIC_PA_TYPES,
 )
 def numeric_pa_type(request):
     return request.param
 
 
+# TODO: Consider adding another fixture/adapting this
+# fixture to consider nullability
+@pytest.fixture(scope="session", params=[0, 100])
+def table_data(request):
+    """
+    Returns (TableWithMetadata, pa_table).
+
+    This is the default fixture you should be using for testing
+    pylibcudf I/O writers.
+
+    Contains one of each category (e.g. int, bool, list, struct)
+    of dtypes.
+    """
+    nrows = request.param
+
+    table_dict = dict()
+    # Colnames in the format expected by
+    # plc.io.TableWithMetadata
+    colnames = []
+
+    for typ in ALL_PA_TYPES:
+        rand_vals = np.random.randint(0, nrows, nrows)
+        child_colnames = []
+
+        if isinstance(typ, pa.ListType):
+
+            def _generate_list_data(typ):
+                child_colnames = []
+                if isinstance(typ, pa.ListType):
+                    # recurse to get vals
+                    rand_arrs, grandchild_colnames = _generate_list_data(
+                        typ.value_type
+                    )
+                    pa_array = pa.array(
+                        [list(row_vals) for row_vals in zip(rand_arrs)],
+                        type=typ,
+                    )
+                    child_colnames.append(("", grandchild_colnames))
+                else:
+                    # typ is scalar type
+                    pa_array = pa.array(rand_vals).cast(typ)
+                    child_colnames.append(("", []))
+                return pa_array, child_colnames
+
+            rand_arr, child_colnames = _generate_list_data(typ)
+        elif isinstance(typ, pa.StructType):
+
+            def _generate_struct_data(typ):
+                child_colnames = []
+                if isinstance(typ, pa.StructType):
+                    # recurse to get vals
+                    rand_arrs = []
+                    for i in range(typ.num_fields):
+                        rand_arr, grandchild_colnames = _generate_struct_data(
+                            typ.field(i).type
+                        )
+                        rand_arrs.append(rand_arr)
+                        child_colnames.append(
+                            (typ.field(i).name, grandchild_colnames)
+                        )
+
+                    pa_array = pa.StructArray.from_arrays(
+                        [rand_arr for rand_arr in rand_arrs],
+                        names=[
+                            typ.field(i).name for i in range(typ.num_fields)
+                        ],
+                    )
+                else:
+                    # typ is scalar type
+                    pa_array = pa.array(rand_vals).cast(typ)
+                return pa_array, child_colnames
+
+            rand_arr, child_colnames = _generate_struct_data(typ)
+        else:
+            rand_arr = pa.array(rand_vals).cast(typ)
```

Comment:
*TODO (perhaps)*: This looks like a wonderful opportunity to use hypothesis to generate the data. i.e. we write rules for production of the base cases and combinators.

That would also allow generation of tests for `list(struct(T))` and `struct(list(T))` (for example).

Reply:
I could have sworn we had an issue about using hypothesis somewhere, but I can't find it. Maybe I never got around to documenting. Yes, I think using it would be great.

Reply:
Created an issue: https://github.com/rapidsai/cudf/issues/16129